wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java

```diff
@@ -24,380 +24,462 @@
-   */
-  public static Float minus(Float x, Float y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x - y;
-  }
-
-  /**
-   * Arithmetic operation - Subtract a column from another column.
-   */
-  public static BigDecimal minus(BigDecimal x, BigDecimal y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x.subtract(y).stripTrailingZeros();
-  }
-
-  /**
-   Arithmetic operation - Multiply a column with another column.
-   */
-  public static Integer multiply(Integer x, Integer y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
-  }
-
-  /**
-   * Arithmetic operation - Multiply a column with another column.
-   */
-  public static Double multiply(Double x, Double y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
+  // The Number types upon which arithmetic operations can be performed, in order of increasing generality
+  private enum NumberType {
+    SHORT, INTEGER, LONG, FLOAT, DOUBLE, BIGDECIMAL
   }
 
   /**
-   * Arithmetic operation - Multiply a column with another column.
+   * Return the "largest" (e.g. most general) type of the given numbers.
+   * Used for operations applied to columns of multiple types, in order to determine the type of the output column.
+   * The order of increasing generality is Short < Integer < Long < Float < Double < BigDecimal.
+   * Returns null if any input is null.
+   * Throws an error if any input of type Byte is given (this is not currently supported for arithmetic ops).
    */
-  public static Float multiply(Float x, Float y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
-  }
-
-  /**
-   * Arithmetic operation - Multiply a column with another column.
-   */
-  public static BigDecimal multiply(BigDecimal x, BigDecimal y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x.multiply(y);
-  }
-
-  /**
-   * Arithmetic operation - Divide a column by another and return the quotient.
-   */
-  public static Integer divideq(Integer x, Integer y) {
-    if (x == null || y == null || y == 0) {
-      return null;
-    }
-    return x / y;
-  }
-
-  /**
-   * Arithmetic operation - Divide a column by another and return the quotient.
-   */
-  public static Double divideq(Double x, Double y) {
-    if (x == null || y == null || y == 0) {
-      return null;
-    }
-    return x / y;
-  }
-
-  /**
-   * Arithmetic operation - Divide a column by another and return the quotient.
-   */
-  public static Float divideq(Float x, Float y) {
-    if (x == null || y == null || y == 0) {
-      return null;
+  private static NumberType getOutputType(Number... nums) throws DirectiveExecutionException {
+    NumberType outputType = NumberType.SHORT;
+    for (Number num : nums) {
+      int outputTypePriority = java.util.Arrays.asList(NumberType.values()).indexOf(outputType);
```

Comment:
Enumerations can have internal "values" attached to them, see https://github.com/cdapio/cdap/blob/163ade54065ecac3a4fd80fe3d76a5aee38474f9/cdap-common/src/main/java/io/cdap/cdap/common/conf/StringUtils.java#L110 as an example.

I think it would be better to have the priority as a value for the enumeration instead of doing an array seek for each input value.

This is an NxM operation, where N is the number of input values and M the length of the array of number type values (so 6)

Reply:
Oo, that's a cool feature I hadn't seen before. Fixed!